### PR TITLE
Adds basic support for IPv6

### PR DESCRIPTION
**Pull request by bnaecker**
- Adds IPv6 support to the OPTE API and main engine types. This includes fleshing out some missing edges for IPv6 addresses and CIDRs, and adding support for specifying IPv6 addresses in router entries, etc. The main type expanded here is the `VpcCfg`, which now supports an `IpCfg` that specifies all L3 information. That supports exactly one IPv4 or IPv6, or one of each, for private addresses. An optional SNAT and external address for each are also supported.
- Updates the `opte-ioctl` and `opteadm` crates to support IPv6, and to use a `VpcCfg` as the argument, rather than a bunch of disparate arguments. Fleshes out handling for IPv6 in router entries, port info and printing, and layer / rule printing.
- Adds a few niceties to the D scripts for pretty-printing IPv6
- Renames a lot of IPv4 specific types, such as `Dhcp4Reply` to `DhcpReply`. Types without a prefix will be assumed to refer to IPv4, and IPv6 will always have a version number.
- Adds an `icmpv6` layer to `opte` and the `oxide-vpc`, and inserts it in the configuration created by the `xde` driver. This supports ICMPv6 echo requests from the guest to the gateway only. An integration test verifying the hairpinned echo reply is also here.
- Updates the API version check script to compare all commits relative to the `master` branch, rather than the last.

### CODE REVIEW

**Comment by bnaecker** [comment]
In addition the tests in the repo, I've also tested we can actually launch guests with Omicron using the updated `xde`. IPv4 support appears unchanged -- guests can still talk to each other and reach the internet. I've also verified that we can `ping6` the gateway from the guest. I have a modified Omicron which hardcodes an address of `fd00:abcd:abcd:abcd::5` in the guest, which I added manually in the guest with `ip -6 addr add`. Also, note that NDP isn't implemented, so you'd need to add the OPTE gateway as a neighbor with: `sudo ip -6 neigh add <IPv6> lladdr a8:40:25:ff:77:77 dev <iface>`. Once that's done, from the guest we see:

```bash
debian@myinst:~$ ping6 fd00:abcd:abcd:abcd::1 -c 1
PING fd00:abcd:abcd:abcd::1(fd00:abcd:abcd:abcd::1) 56 data bytes
64 bytes from fd00:abcd:abcd:abcd::1: icmp_seq=1 ttl=255 time=0.126 ms

--- fd00:abcd:abcd:abcd::1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.126/0.126/0.126/0.000 ms
debian@myinst:~$
```

OPTE is actually hairpinning the packet:

```bash
bnaecker@feldspar : ~/opte $ pfexec ./dtrace/opte-trace opte-rule-match
<SNIP>
PORT     LAYER        MATCH  DIR FLOW                                        ACTION
opte0    icmpv6       YES    out ICMPv6,[fd00:abcd:abcd:abcd::5]:0,[fd00:abcd:abcd:abcd::1]:0 HAIRPIN: ICMPv6 Echo Reply (A8:40:25:FF:77:77,fd00:abcd:abcd:abcd::1) => (A8:40:25:FA:4E:42,fd00:abcd:abcd:abcd::5)
^C
```

File: oxide-vpc/tests/integration_tests.rs
Comment (rzezeski) [inline oxide-vpc/tests/integration_tests.rs:501]: Someone please notify past-Ryan that all IPs are routable, just a matter of how far that route can traverse. This makes me realize that `public_ip` was a poor choice I made long ago. We should probably rename this `external_ip` and remove this comment in the test.

File: oxide-vpc/tests/integration_tests.rs
Comment (rzezeski) [inline oxide-vpc/tests/integration_tests.rs:568]: I don't think you meant to give guest2 the same mac as guest1.

**Review by rzezeski — COMMENTED** [review]
This is a lot of great work. It feels good to see various IPv6 `todo!`s eliminated! I think most of my comments are pretty straightforward and shouldn't require much modification. This is pretty much how I would have done all this work; though I don't think I would have thought of some of the nice simplifications you made in various places.

File: opte-ioctl/src/lib.rs
Comment (bnaecker) [reply opte-ioctl/src/lib.rs:7]: I have no idea how that got there, thanks for catching!

File: opte/src/engine/ip6.rs
Comment (bnaecker) [reply opte/src/engine/ip6.rs:57]: Same answer as for `Ipv4Meta`, it's just a habit to make a type `Copy` when that's possible. Out of curiosity, why did you leave it off these types?

File: opte/src/engine/ip6.rs
Comment (bnaecker) [reply opte/src/engine/ip6.rs:265]: Done.

File: opte/src/engine/nat.rs
Comment (bnaecker) [reply opte/src/engine/nat.rs:33]: Same as above, depending on how we resolve that, I'll do the same thing.

File: opte/src/engine/nat.rs
Comment (bnaecker) [reply opte/src/engine/nat.rs:168]: Yeah, I started going down that route and backed off. I think I also changed which types accept `Ipv4Addr` and which take `IpAddr`, though that doesn't seem to apply to this test. Would you prefer I just revert all the changes in this test function?

File: opte/src/engine/snat.rs
Comment (bnaecker) [reply opte/src/engine/snat.rs:47]: Yeah, I didn't have a great name in mind. What you suggest is better, I'll use one of those. Thanks!

File: opte/src/engine/snat.rs
Comment (bnaecker) [reply opte/src/engine/snat.rs:344]: Fair enough, I can remove it. This was part of handling the change to `SNat` to take an `IpAddr` of either version, but I agree it's pretty clear here.

File: oxide-vpc/src/api.rs
Comment (bnaecker) [reply oxide-vpc/src/api.rs:31]: I think my (implicit) rule is to leave them off unless there are multiple sentences. That's kinda silly though, I'll make a pass and add them.

File: oxide-vpc/src/api.rs
Comment (bnaecker) [reply oxide-vpc/src/api.rs:29]: Similar to the question above, why is that?

File: oxide-vpc/src/api.rs
Comment (bnaecker) [reply oxide-vpc/src/api.rs:99]: Will do.

File: oxide-vpc/src/engine/arp.rs
Comment (bnaecker) [reply oxide-vpc/src/engine/arp.rs:35]: Ah, good catch thanks. I'll rework this.

Comment (bnaecker) [reply oxide-vpc/src/engine/arp.rs:35]: I've changed this to conditionally respond to ARP requests when there is an IPv4 address, and always drop other ARP requests in either direction. I'm going to test this with a guest now, and I'll resolve this after I verify the behavior.

File: oxide-vpc/src/engine/dhcp.rs
Comment (bnaecker) [reply oxide-vpc/src/engine/dhcp.rs:45]: The idea of a "gateway" layer that acts that way is definitely appealing. It centralizes the logic for deciding which packets make it anywhere beyond the guest. At the same time, having layers for each protocol has definitely been helpful, for testing and development. It's nice to be able to see just ICMP or whatever, for example when using `opteadm dump-layer`.

I've been thinking that it would be very nice to name rules within layers, even before you bring this up. This would I think make that more valuable, since one could read a name for the rule, say `icmpv6-guest-echo-reply`, and see the actions and predicates associated with that. It gets a bit funky with rules that are related, like the inbound and outbound rules to drop ARP, but I still think it might be helpful.

File: oxide-vpc/src/engine/icmpv6.rs
Comment (bnaecker) [reply oxide-vpc/src/engine/icmpv6.rs:49]: The intention was to separate them so that adding the NDP stuff didn't clutter the ICMP layer for version 4. But I agree that's just organization, and shouldn't affect the behavior. But it does relate to my above idea of naming rules. The value there would be to look at the output of `opteadm dump-layer icmp`, and see the NDP-related rules or ICMP version 4 rules. Right now, one needs to basically decode the predicates to know what rule is supposed to be doing.

File: oxide-vpc/src/engine/router.rs
Comment (bnaecker) [reply oxide-vpc/src/engine/router.rs:168]: Thanks for catching -- I meant to come back and fix that up. I'll improve that error now.

File: oxide-vpc/src/engine/router.rs
Comment (bnaecker) [reply oxide-vpc/src/engine/router.rs:150]: I see your point though, I think it'd be nice to put the default check inside that one method to centralize the "validity checking" bits. I'll do that now.

As for the larger question, I don't think there's anything in RFD 21 that _precludes_ users from adding a route for any IP address or VPC Subnet to the IG. I'll leave this as-is for now, other than moving the check for `is_default` into the function checking validity of the dest/target pair.

File: xde/src/xde.rs
Comment (bnaecker) [reply xde/src/xde.rs:472]: What I meant was that the mapping for a guest IPv4 and IPv6 address should return the same `Virt2Phys` object, which refers to the IPv6 address of just one host. I.e., both the IPv4 and IPv6 addresses "live" on the same host. Does that make sense? Or am I missing something? If it does, I'll add some more detail since it's clearly confusing.

File: xde/src/xde.rs
Comment (bnaecker) [reply xde/src/xde.rs:1444]: I think that's because `cargo fmt` is pretty conservative about modifying anything in macros. I'll fix it up manually, thanks.

File: oxide-vpc/tests/integration_tests.rs
Comment (bnaecker) [reply oxide-vpc/tests/integration_tests.rs:501]: Renaming "public" to "external" seems like a good idea. Would you like me to do that now, or make another issue? In either case, I'll remove the comment.

File: oxide-vpc/tests/integration_tests.rs
Comment (bnaecker) [reply oxide-vpc/tests/integration_tests.rs:568]: I did not, no. That looks like a copy-paste error, thanks for catching!

File: opte-api/src/ip.rs
Comment (bnaecker) [reply opte-api/src/ip.rs:341]: Thanks, that's a good point about the error. I can make it something version-agnostic.

But I'm not sure I agree with the part that we always know which version we're parsing from the command line. In this PR, I changed the `opteadm add-router-entry-ipv4` command to just `add-router-entry`. That accepts either an IPv4 or IPv6 address or CIDR block, which is exactly why I added this parsing code. It seemed better to simplify the CLI and the underlying `ioctl(2)` interface to accept an `IpAddr` rather than have nearly identical versions.

That said, I get the point -- it's exactly why we carry the IP version "up" through types like `Ipv4Cfg` and `Ipv6Cfg`, so that we can be sure at compile-time that the actual versions are the same for all data inside them. By _not_ doing that here, I had to add the runtime checks in the `valid_router_dest_target_pair` function, which checks that the destination and target are of the same protocol version. If you'd prefer, I can go the other way: keep IPv4 and IPv6 versions of the router APIs, and remove the runtime checks.

Comment (rzezeski) [reply opte-api/src/ip.rs]: Yea, and I hadn't even thought of the fact that we are pushing off dest+target validation to runtime. It seems like it could be nice to eliminate this runtime check from the router entry API. I haven't thought this through, but could we just give AddRouterEntryReq the Ip4/Ip6 enum treatment, and then the command line can either have two commands or one command with an option to indicate type.

That said, this is all nit picky and could be done in a follow up if we agree it would be cleaner this way. I don't want to thrash this PR too much. So I'll leave this up to you.

**Comment by bnaecker** [comment]
The `opte-xde` check is failing on a format-check, though I really don't like how `cargo fmt` wants to format the block. This might be some interaction between a line-length limit and the wrapping logic, but this in particular is horrible:

```rust
1437                         } else {
1438                             // This should be unreachable!(). We have an IP header,
1439                             // but neither `ip4()` nor `ip6()` returned something.
1440                             opte::engine::dbg(format!(
1441                             "Packet contains IP header, but neither `ip4()` \
1442                                 nor `ip6()` returned `Some`. IP header: {:?}",
1443                             ip_hdr,
1444                         ));
1445                             false
1446
```

That's putting the closing parenthesis for the function call at the same indention as the surrounding `if-else` bracing, which is extremely confusing. I'm going to play with the line length a bit, see if I can't convince `cargo fmt` to...not suck.

File: opte/src/engine/ip4.rs
Comment (rzezeski) [reply opte/src/engine/ip4.rs:174]: At some point I forgot Copy is really just about move semantics and got my wires crossed to associate it with bcopy'ing, even when looking at a reference -- which is totally not how this works at all, haha. You are right, no reason these types shouldn't be Copy.

File: opte/src/engine/nat.rs
Comment (rzezeski) [reply opte/src/engine/nat.rs:168]: I won't die if you leave it, but truth is if I have to modify this test in the future I might change them back, haha. Up to you.

File: oxide-vpc/src/engine/icmpv6.rs
Comment (rzezeski) [reply oxide-vpc/src/engine/icmpv6.rs:49]: Hmmmm, that's a good point. If you haven't already gone done that path of refactoring this, then maybe hold off. We can always reorg later, and you make a good point about the naming of rules.

File: xde/src/xde.rs
Comment (rzezeski) [reply xde/src/xde.rs:472]: Oh duh. Sorry I just had a brain fart.

File: oxide-vpc/tests/integration_tests.rs
Comment (rzezeski) [reply oxide-vpc/tests/integration_tests.rs]: I'll leave it up to you on doing it in this PR vs. follow up.

**Comment by rzezeski** [comment]
No worries on cargo fmt, it is what it is, I wouldn't spent too much time trying to get it to heel.

File: opte/src/engine/ip4.rs
Comment (bnaecker) [reply opte/src/engine/ip4.rs:174]: Ah, ok that makes sense. Yeah, it's just indicating that moving out of a value means the original can still be referenced, since it's implicitly copied. I'll put the derive in order now, and then mark these and the places resolved. Thanks!

File: oxide-vpc/src/engine/arp.rs
Comment (rzezeski) [inline oxide-vpc/src/engine/arp.rs:66]: Leftovers.

Comment (rzezeski) [reply oxide-vpc/src/engine/arp.rs:66]: Oh wait, this is from me. LOL. This is related to ext-ip-hack. Nevermind.

File: opte/src/engine/nat.rs
Comment (bnaecker) [reply opte/src/engine/nat.rs:168]: Reverted, no need to make your merges a pain. There are still some changes there, to handle the renaming `Nat4` -> `Nat`, but that should be it now.

File: oxide-vpc/tests/integration_tests.rs
Comment (bnaecker) [reply oxide-vpc/tests/integration_tests.rs]: We'll do it live. I'll change these now, while there's a ton of API churn anyway.

File: oxide-vpc/src/engine/arp.rs
Comment (bnaecker) [reply oxide-vpc/src/engine/arp.rs]: @rzezeski indicated offline that the new version of this layer-building looks good, but I'll leave this open for explicit approval from him.

File: oxide-vpc/src/engine/icmpv6.rs
Comment (bnaecker) [reply oxide-vpc/src/engine/icmpv6.rs:49]: Sounds good. I'll leave the ICMP and ICMPv6 layers separate for now. When I get through NDP we can reevaluate.

**Comment by bnaecker** [comment]
Got the fmt resolved. It was line length, I just split the lines into shorter pieces and it did the right thing here.

**Review by rzezeski — APPROVED** [review]

File: opte-api/src/ip.rs
Comment (bnaecker) [reply opte-api/src/ip.rs]: I think I'm going to defer this. Making the `AddRouterEntryReq` into a V4 and V6 version means we also need to make the `RouterTarget` into the same. That's kinda lame, because the `Drop` and `InternetGateway` variants would be duplicated. I think the runtime check isn't great either, but it's not obvious how to make that much better without a bunch of other code.

Comment (rzezeski) [reply opte-api/src/ip.rs]: Okay, that's fine. I definitely don't want to hold this up.

Comment (bnaecker) [reply opte-api/src/ip.rs]: Awesome, thanks. I'll make an issue referring to this code once I merge. Running the last test to verify the ARP layer changes in a guest now.

Comment (bnaecker) [reply opte-api/src/ip.rs]: Ok, I've checked that guests behave the same way now, and that the right traffic is dropped (Inbound ARP other than the Proxy ARP requests, outbound other than for the gateway).

**Comment by bnaecker** [comment]
Thanks for the detailed @rzezeski!

**Pull request merged**
